axum-handle-error-extract/src/lib.rs

```diff
@@ -0,0 +1,389 @@
+    missing_docs
+)]
+#![deny(unreachable_pub, private_in_public)]
+#![allow(elided_lifetimes_in_paths, clippy::type_complexity)]
+#![forbid(unsafe_code)]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(test, allow(clippy::float_cmp))]
+
+use axum::{
+    body::{box_body, BoxBody, Bytes, Full, HttpBody},
+    extract::{FromRequest, RequestParts},
+    http::{Request, Response, StatusCode},
+    response::IntoResponse,
+    BoxError,
+};
+use pin_project_lite::pin_project;
+use std::{
+    convert::Infallible,
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::ServiceExt;
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// [`Layer`] that applies [`HandleError`] which is a [`Service`] adapter
+/// that handles errors by converting them into responses.
+///
+/// See [module docs](self) for more details on axum's error handling model.
+pub struct HandleErrorLayer<F, T> {
+    f: F,
+    _extractor: PhantomData<fn() -> T>,
+}
+
+impl<F, T> HandleErrorLayer<F, T> {
+    /// Create a new `HandleErrorLayer`.
+    pub fn new(f: F) -> Self {
+        Self {
+            f,
+            _extractor: PhantomData,
+        }
+    }
+}
+
+impl<F, T> Clone for HandleErrorLayer<F, T>
+where
+    F: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            f: self.f.clone(),
+            _extractor: PhantomData,
+        }
+    }
+}
+
+impl<F, E> fmt::Debug for HandleErrorLayer<F, E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("HandleErrorLayer")
+            .field("f", &format_args!("{}", std::any::type_name::<F>()))
+            .finish()
+    }
+}
+
+impl<S, F, T> Layer<S> for HandleErrorLayer<F, T>
+where
+    F: Clone,
+{
+    type Service = HandleError<S, F, T>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        HandleError::new(inner, self.f.clone())
+    }
+}
+
+/// A [`Service`] adapter that handles errors by converting them into responses.
+///
+/// See [module docs](self) for more details on axum's error handling model.
+pub struct HandleError<S, F, T> {
+    inner: S,
+    f: F,
+    _extractor: PhantomData<fn() -> T>,
+}
+
+impl<S, F, T> HandleError<S, F, T> {
+    /// Create a new `HandleError`.
+    pub fn new(inner: S, f: F) -> Self {
+        Self {
+            inner,
+            f,
+            _extractor: PhantomData,
+        }
+    }
+}
+
+impl<S, F, T> Clone for HandleError<S, F, T>
+where
+    S: Clone,
+    F: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            f: self.f.clone(),
+            _extractor: PhantomData,
+        }
+    }
+}
+
+impl<S, F, E> fmt::Debug for HandleError<S, F, E>
+where
+    S: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("HandleError")
+            .field("inner", &self.inner)
+            .field("f", &format_args!("{}", std::any::type_name::<F>()))
```

Comment:
Wouldn't `&type_name()`, w/o the `format_args!`, work too?

Reply:
Yeah I think thats right! No idea why we wrote it like this in tower. Seems to do the same.

Reply:
I think I know now actually 😅

W/o `format_args` it will add quotes, right?

Reply:
Oh yeah thats true. I'll revert to keep the `format_args`.